third_party/blink/renderer/core/record_replay/lifecycle.cc

```diff
@@ -0,0 +1,135 @@
+
+#include "base/record_replay_paint_surface.h"
+#include "third_party/blink/renderer/core/frame/local_frame.h"
+#include "third_party/blink/renderer/core/page/focus_controller.h"
+#include "third_party/blink/renderer/core/page/page.h"
+#include "third_party/blink/renderer/core/record_replay/lifecycle.h"
+#include "third_party/blink/renderer/platform/wtf/casting.h"
+
+#define LOG_LIFECYCLE_EVENTS 1
+
+namespace recordreplay {
+
+// the current set of ordinary pages, presumably scoped to this renderer process
+// (TODO verify this last part) lives at blink::Page::OrdinaryPages().  We
+// should be able to use this to get the JS contexts for each main frame, but I
+// haven't traced through.
+
+// There might be a reference to our last-active page someplace in blink
+// already, but until/unless we find it, we'll update it here.  It should be
+// safe to keep the raw pointer to the page, since we'll clear it when the page
+// is destroyed.
+static blink::Page* last_active_page = nullptr;
+
+static void set_last_active_page(blink::Page* page) {
+    if (page == last_active_page) {
+        // it's already the active page
+        return;
+    }
+
+#if LOG_LIFECYCLE_EVENTS
+    recordreplay::Print("  - updating last-active page\n");
```

Comment:
1. We don't need `\n`. This will call `PrintJSON`, which outputs a JSON object followed by `\n`.
2. Can we give life cycle events a "[Label]"? E.g. `Print("[ContextRoots] ...")` or `Print("[Roots]")?

Note: I would use the label `Roots` or `ContextRoots`, since these events ultimately affect our tracked root data structures, either the live set or the last active root.

Reply:
went with [ContextRoots]